### PR TITLE
Update FileStorage's docs and tests

### PR DESCRIPTION
**Pull request by wiryls**
[GSoC 2016] FileStorage.

Update Base64 functions and also add some test and documentation.

### This pullrequest changes

 - Make `cvWriteRawDataBase64` compatible with `cvWriteRawData` and so on.
 - Add a Base64 flag for FileStorage to output data in Base64 automatically.
 - Make one test of ml module (`opencv/modules/ml/test/test_save_load.cpp`) use Base64.
 - Add some accuracy testing for Base64.
 - Add a basic performance test for Base64.
 - Add documentation.

### CODE REVIEW

File: modules/core/include/opencv2/core/core_c.h [hunk 1]
```diff
@@ -1976,15 +1976,27 @@ CVAPI(void) cvSetIPLAllocators( Cv_iplCreateImageHeader create_header,
...
-CvFileStorage structure. If the file cannot be opened then the function returns NULL.
+filename extension: .xml for XML and .yml or .yaml for YAML.
+
+At the same time, it also supports adding parameters like "example.xml?base64".
+@code
+    CvFileStorage* fs = cvOpenFileStorage( "example.yml?base64", 0, CV_STORAGE_WRITE );
+@endcode
+it's exactly the same as
+@code
+    CvFileStorage* fs = cvOpenFileStorage( "example.yml", 0, CV_STORAGE_WRITE_BASE64 );
```
Comment (ilya-lavrenov) [inline modules/core/include/opencv2/core/core_c.h]: see https://github.com/opencv/opencv/issues/6938. Please, extract code snippets to a separate cpp file.

Comment (wiryls) [inline modules/core/include/opencv2/core/core_c.h]: @ilya-lavrenov 
OK, and should I create a new sample file and put it in `samples/cpp` folder?
If so, are there any rules for the file path?

File: modules/core/src/persistence.cpp [hunk 2]
```diff
@@ -1031,6 +1064,188 @@ static double icv_strtod( CvFileStorage* fs, char* ptr, char** endptr )
...
+    fs->state_of_writing_base64 = state;
+}
+
+
+static void check_if_write_struct_is_delayed( CvFileStorage* fs, bool change_type_to_base64 = false )
+{
+    if ( fs->is_write_struct_delayed )
+    {
+        /* save data to prevent recursive call errors */
+        char * struct_key = 0;
```
Comment (vpisarev) [inline modules/core/src/persistence.cpp]: please, use std::string instead of manually allocated `char*`

File: modules/core/src/persistence.cpp [hunk 3]
```diff
@@ -1119,50 +1334,54 @@ static char* icvYMLParseBase64(CvFileStorage* fs, char* ptr, int indent, CvFileN
...
             CV_PARSE_ERROR("Cannot parse dt in Base64 header");
 
         beg += base64::ENCODED_HEADER_SIZE;
     }
 
-    /* buffer for decoded data(exclude header) */
-    std::vector<uchar> buffer(total_byte_size + 4);
+    /* get all Base64 data */
+    std::string base64_buffer;
+    base64_buffer.reserve( 16U * 1024U * 1024U );
```
Comment (vpisarev) [inline modules/core/src/persistence.cpp]: Let's reserve less than that initially, say, 1MB

File: modules/core/src/persistence.cpp [hunk 4]
```diff
@@ -243,6 +256,13 @@ typedef struct CvFileStorage
     std::deque<char>* outbuf;
 
     base64::Base64Writer * base64_writer;
+    bool is_default_using_base64;
+    base64::fs::State state_of_writing_base64;  /**< used in WriteRawData only */
+
+    bool is_write_struct_delayed;
+    char* delayed_struct_key;
```
Comment (vpisarev) [inline modules/core/src/persistence.cpp]: let's use std::string for these things too.
use std::string::empty() to check whether the strings are empty

File: modules/core/src/persistence.cpp [hunk 5]
```diff
@@ -243,6 +256,13 @@ typedef struct CvFileStorage
...
+    int   delayed_struct_flags;
+    char* delayed_type_name;
```
Comment (vpisarev) [inline modules/core/src/persistence.cpp]: same

File: modules/core/src/persistence.cpp (hunk 4 above)
Comment (wiryls) [inline modules/core/src/persistence.cpp]: @vpisarev, But `struct CvFileStorage` is a POD. Using The STL string may change it and cause some trouble. e.g. I remember `CvFileStorage` will be initialized via `memset`. But STL string cannot be initialized like this.

Comment (vpisarev) [inline modules/core/src/persistence.cpp]: @wiryls, you are right, this is a valid point

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**